### DISCUSSION TITLE
Programatically move a note

### DISCUSSION THREAD

**Discussion by jobidon** [TriliumNext/Trilium#2507] [Q&A] [answered]
I am trying to move a note in the tree
I found the toggleNoteInParent function which allow me to attach a note to the new path, that's fine, but I am trying to move the note, so I need to remove the note from its current branch as well.
I am assuming that the getParentNotes, getAncestors, getParentBranches functions would consider the clones/copies of the note in the tree, so what I really need is the parent of originEntity.getNote(). How do I obtain that? 
Or is there an easier way to move a note in the tree?

**Comment by zadam** [comment, marked as answer]
Hi, what about this (roughly)?

```javascript
const note = ... // note I want to move
const originalParentNoteId = note.getParentNotes()[0].parentNoteId;

api.toggleInParent(true, note.noteId, newParentNoteId); // clone to new location
api.toggleInParent(false, note.noteId, originalParentNoteId); // remove the old location
```

**Comment by nicollasricas** [comment]
I'm also trying to move a note programmatically but couldn't find any way.

```javascript
module.exports = async function (noteId) {
  const note = await api.getNote(noteId);

  if (!note) {
    return;
  }

  if (note.getParentNotes().some((m) => m.hasLabel("todoArchive"))) {
    return;
  }

  const todoDoneId = (
    await note.getParentNotes()[0].getParentNotes()[0].getChildNotes()
  ).find((m) => m.hasLabel("todoDone"))?.noteId;

  if (!todoDoneId) {
    return;
  }

  note.removeLabel("timeLeft");

  await api.toggleNoteInParent(false, note.noteId, todoDoneId);
};
```

If I set the first argument to `true`, it works but clones the note instead of moving it.

Here's what I'm trying to do: move this "feature note" to the "done" section.

![image](https://user-images.githubusercontent.com/7860985/173257523-9853e651-94a5-45fa-b6e7-64b3f7a3b622.png)

**Reply by zadam** [reply]
Hi, yes, this is expected.

Trilium doesn't have a "move note" API method, because it doesn't play conceptually well. All is good as long as the note has a single parent, but what does "move note" mean when there are 10 parents? What's the expected state after calling "move" on such note?

**Reply by nicollasricas** [reply]
There is no such thing as multiple parents. A note can have only one parent period, and as long I set a new parent to be the child, I think it should be possible.

A
||||||||| B
|||||||||||||||||||| C

Move to A? Fine, it will be a child of A with B as "brother".
Move to B? Go for it. It will be a child of B with C as "brother".
Move to C? Have it a go, my friend. You are the only child of C congrats!

Another way to solve this (hack) would be cloning it (copying) and removing the clone link? I don't know if that is possible.

**Reply by nicollasricas** [reply]
I mean, "there's no such thing as multiple parents" because trillium handles the tree (visually) where every note is a child of another note except for the root notes. Beyond that point, everyone else is a child of a single parent.

**Reply by sigaloid** [reply]
@nicollasricas: Yes, it is possible to have multiple parents. Please read [this wiki page](https://github.com/zadam/trilium/wiki/Cloning-notes#solution) about note cloning:

>The solution to the problem shown above is to allow notes to have multiple parents.

>I call these "clones", but that is a bit misleading - there's no original and cloned note - the notes in both of the parents categories are identical.

**Reply by nicollasricas** [reply]
Thanks for the help @sigaloid, but I'm trying to move a note and not cloning it.

I will take the longer path and create a new note inside the parent that I wish to move, copy everything and delete the original. I was just looking for a easier way!

**Reply by sigaloid** [reply]
I believe that is what zadam describes in the solution to this discussion (correct me if I'm wrong): https://github.com/zadam/trilium/discussions/2507#discussioncomment-1907930

**Reply by nicollasricas** [reply]
F me, I found out why it was always cloning the note. I had a previous attribute configured for testing the new update @zadam made. It was cloning after moving it. I didn't see that and was thinking that the "roughly move" wasn't working.

Thanks for all the help!

**Reply by nicollasricas** [reply]
@sigaloid. Here's why I was thinking the note were cloned and not moved.

![image](https://user-images.githubusercontent.com/7860985/173453244-281d1f8c-9af8-448e-bbfe-f22b90a71c0b.png)

Backend logs:
```
18:05:48.631 [Script TyCL5FFAfV6H] Ensured note bcfcRjT7m5M4 is in parent note iknHvZF0hlRh with prefix undefined
18:05:48.632 [Script TyCL5FFAfV6H] Marking branches 51Kin9H82Ybe as deleted
18:05:48.633 [Script TyCL5FFAfV6H] Ensured note bcfcRjT7m5M4 is NOT in parent note cLYeDipjZqYh
```

After reloading the frontend:

![image](https://user-images.githubusercontent.com/7860985/173453687-3b7dc720-6783-4702-b2a8-86d3c6471c82.png)


I hope this helps someone in the future with the same issue.

@zadam Do you see any downside of calling refreshTree or reloadNote after moving the note?

**Reply by zadam** [reply]
`refreshTree()` is a no-op (tree self updates, kept in API for BC), `reloadNote()` shouldn't have negative effects, but also shouldn't be needed.
